### PR TITLE
132 channel sub unsub update

### PR DESCRIPTION
**Pull request by anagha-ks**
#### Description
For SubPayChUpdates, UnsubPayChUpdates API:
- Update it to return the newly defined error type (#132).
- Log the request, error/success response are logged in a structured
format (#133).

##### Category

Implementation Task

##### Relevant issue
Relates to #132, #133

#### Testing

1. Updated `Test_SubPayChUpdates` & `Test_UnsubPayChUpdates` in `payment` package
2. Updated `Test_SubUnsubChUpdate` in `session` package

##### Steps to run the tests

1.All existing tests should pass
2.
3.  

#### Checklist 

- [x] Name is added to the NOTICE file, if it is not present already.
- [x] Changes are rebased onto the target branch.

### CODE REVIEW

**Comment by anagha-ks** [comment]
Rebased onto current develop.

File: session/channel.go [hunk 1]
```diff
@@ -226,18 +226,22 @@ func makeChUpdateNotif(currChInfo perun.ChInfo, proposedState *pchannel.State, e
 }
 
 // SubChUpdates implements chAPI.SubChUpdates.
-func (ch *Channel) SubChUpdates(notifier perun.ChUpdateNotifier) error {
+func (ch *Channel) SubChUpdates(notifier perun.ChUpdateNotifier) perun.APIErrorV2 {
 	ch.Debug("Received request: channel.SubChUpdates")
```
Comment (manoranjith) [inline session/channel.go]: Can you also update the log message to include the method name as a field. You can use your previous PR  (#153) as reference.

File: session/channel.go [hunk 2]
```diff
@@ -226,18 +226,22 @@ func makeChUpdateNotif(currChInfo perun.ChInfo, proposedState *pchannel.State, e
...
+		ch.WithFields(perun.APIErrV2AsMap("SubChUpdates", apiErr)).Error(apiErr.Message())
+		return apiErr
 	}
 
 	if ch.chUpdateNotifier != nil {
-		ch.Error(perun.ErrSubAlreadyExists)
-		return perun.ErrSubAlreadyExists
+		apiErr := perun.NewAPIErrV2ResourceExists("subscription to channel updates", ch.ID(), ErrSubAlreadyExists.Error())
+		ch.WithFields(perun.APIErrV2AsMap("SubChUpdates", apiErr)).Error(apiErr.Message())
+		ch.Error(apiErr)
```
Comment (manoranjith) [inline session/channel.go]:
```suggestion
```
The error is already logged in the previous line.

File: session/channel.go [hunk 3]
```diff
@@ -250,18 +254,21 @@ func (ch *Channel) SubChUpdates(notifier perun.ChUpdateNotifier) error {
 }
 
 // UnsubChUpdates implements chAPI.UnsubChUpdates.
-func (ch *Channel) UnsubChUpdates() error {
+func (ch *Channel) UnsubChUpdates() perun.APIErrorV2 {
 	ch.Debug("Received request: channel.UnsubChUpdates")
```
Comment (manoranjith) [inline session/channel.go]: Update the log message. See the above [comment](https://github.com/hyperledger-labs/perun-node/pull/157/files#r582712783).

File: session/channel.go [hunk 4]
```diff
@@ -250,18 +254,21 @@ func (ch *Channel) SubChUpdates(notifier perun.ChUpdateNotifier) error {
...
-		return perun.ErrChClosed
+		apiErr := perun.NewAPIErrV2FailedPreCondition(ErrChClosed.Error())
+		ch.WithFields(perun.APIErrV2AsMap("UnsubChUpdates", apiErr)).Error(apiErr.Message())
+		return apiErr
 	}
 
 	if ch.chUpdateNotifier == nil {
-		ch.Error(perun.ErrNoActiveSub)
-		return perun.ErrNoActiveSub
+		apiErr := perun.NewAPIErrV2ResourceNotFound("unsubscription to channel updates", ch.ID(), ErrNoActiveSub.Error())
```
Comment (manoranjith) [inline session/channel.go]:
```suggestion
		apiErr := perun.NewAPIErrV2ResourceNotFound("subscription to channel updates", ch.ID(), ErrNoActiveSub.Error())
```

It is `subscription to channel updates` that is not found.

File: session/channel_test.go [hunk 5]
```diff
@@ -192,6 +201,15 @@ func Test_SubUnsubChUpdate(t *testing.T) {
 	err = ch.UnsubChUpdates()
 	require.Error(t, err)
 
+	wantMessage = session.ErrNoActiveSub.Error()
+	assert.Equal(t, perun.ClientError, err.Category())
+	assert.Equal(t, perun.ErrV2ResourceNotFound, err.Code())
+	assert.Contains(t, err.Message(), wantMessage)
+	addInfo1, ok := err.AddInfo().(perun.ErrV2InfoResourceNotFound)
+	require.True(t, ok)
+	assert.Equal(t, "unsubscription to channel updates", addInfo1.Type)
```
Comment (manoranjith) [inline session/channel_test.go]:
```suggestion
	assert.Equal(t, "subscription to channel updates", addInfo1.Type)
```

Comment (manoranjith) [reply session/channel_test.go]: Could you also add assertions on the returned error for the sub tests `Sub_channelClosed` and `Unsub_channelClosed`.

File: session/role_integ_test.go [hunk 6]
```diff
@@ -329,7 +329,9 @@ func Test_Integ_Role(t *testing.T) {
 		err = bobCh.UnsubChUpdates()
 		assert.Error(t, err, "bob unsubscribing from channel updates after closing notification should error")
 		t.Log(err)
-		require.EqualError(t, err, perun.ErrChClosed.Error())
+
+		apiError := perun.NewAPIErrV2FailedPreCondition(session.ErrChClosed.Error())
+		require.Equal(t, err, apiError)
```
Comment (manoranjith) [inline session/role_integ_test.go:334]: Assert for the individual properties of the error as it has been done in other parts of this PR.

File: session/session.go [hunk 7]
```diff
@@ -667,6 +669,28 @@ func (s *Session) GetChsInfo() []perun.ChInfo {
...
+// GetChV2 is a wrapper over GetCh that returns error in the
+// newly defined APIErrorV2 format introduced for the purpose of refactoring.
+//
+// See doc comments on SessionAPI interface for more details.
+// TODO: merge this with GetCh api once GetChV2 is removed from SessionAPI.
+func (s *Session) GetChV2(chID string) (perun.ChAPI, perun.APIErrorV2) {
+	s.WithField("method", "GetChV2").Info("Received request with params:", chID)
+	ch, err := s.GetCh(chID)
+	var apiErr perun.APIErrorV2
+
```
Comment (manoranjith) [inline session/session.go]:
```suggestion
```

Unnecessary new line.

**Review by manoranjith — CHANGES_REQUESTED** [review]

**Review by manoranjith — APPROVED** [review]

**Pull request merged**